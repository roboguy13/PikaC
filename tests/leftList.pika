%generate leftList

Sll : layout[-x](List);
Sll Nil := emp;
Sll (Cons head tail) :=
  exists +nxt.
  x :-> head **
  (x+1) :-> nxt **
  Sll tail [nxt];

TreeLayout : layout[-x](Tree);
TreeLayout Leaf := emp;
TreeLayout (Node payload left right) :=
  exists +p, +q.
  x :-> payload **
  (x+1) :-> p **
  (x+2) :-> q **
  TreeLayout left [p] **
  TreeLayout right [q];

leftList : TreeLayout -> Sll;
leftList Leaf := Nil [Sll];
leftList (Node a b c) :=
  Cons a (leftList b) [Sll];

-- %test "leftList" Sll: leftList (Node 1 (Node 2 (Leaf [TreeLayout]) (Leaf [TreeLayout]) [TreeLayout]) (Leaf [TreeLayout]) [TreeLayout]);
%test "leftList" Sll: leftList (Node 1 (Node 2 (Leaf [TreeLayout]) (Leaf [TreeLayout]) [TreeLayout]) (Node 3 (Leaf [TreeLayout]) (Leaf [TreeLayout]) [TreeLayout]) [TreeLayout]);

