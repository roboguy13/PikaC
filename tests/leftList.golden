void leftList(loc x, loc x1);
void leftList(loc x, loc x1) {
 if (x == 0) {
  WRITE_LOC(x1, 0, 0);
 } else {
  loc pa = READ_LOC(x, 0);
  loc p = READ_LOC(x, 1);
  leftList(p, x1);
  loc x3 = READ_LOC(x1, 0);
  loc x = (loc)malloc(2 * sizeof(loc));
  WRITE_LOC(x1, 0, x);
  WRITE_LOC(x, 1, x3);
  WRITE_LOC(x, 0, pa);
 }
}